### PR TITLE
Improve resilience of secret file management. Fixes #2093

### LINKED ISSUE #2093: Improve resilience of secret file management
When creating secret files, the runtime needs to ensure that it doesn't attempt to overwrite existing secret files and that inability to create a file (IO exceptions due to another instance or Kudu having a write lock to that file) are gracefully handled.

### PR DESCRIPTION
**Pull request by alrod**

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Security/SecretManager.cs [hunk 1]
```diff
@@ -446,5 +425,76 @@ private void OnSecretsChanged(object sender, SecretsChangedEventArgs e)
...
+                    if (hostSecrets == null)
+                    {
+                        _traceWriter.Verbose(Resources.TraceHostSecretGeneration);
+                        _logger?.LogDebug(Resources.TraceHostSecretGeneration);
+                        hostSecrets = GenerateHostSecrets();
+                        await PersistSecretsAsync(hostSecrets);
+                    }
+                    return hostSecrets;
+                }
+                catch (Exception ex)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/SecretManager.cs]: If we switch this to catch the more specific `IOException`, we'll only fall into our catch if we match the exception type and avoid doing so when it isn't. That also allows us to remove the `IsTransient` method as it is the only thing it's checking for.

You can also use an exception filter, which would improve this (using the `when` clause)

File: src/WebJobs.Script.WebHost/Security/SecretManager.cs [hunk 2]
```diff
@@ -446,5 +425,76 @@ private void OnSecretsChanged(object sender, SecretsChangedEventArgs e)
 
             await _repository.PurgeOldSecretsAsync(currentFunctions, traceWriter, logger);
         }
+
+        private async Task<HostSecrets> LoadHostSecretsWithRetryAsync()
+        {
+            int currentRetry = 0;
+            for (; ;)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/SecretManager.cs]: With the exception filter mentioned below, you can define and increment your counter in your `for` statement

File: src/WebJobs.Script.WebHost/Security/SecretManager.cs [hunk 3]
```diff
@@ -446,5 +425,76 @@ private void OnSecretsChanged(object sender, SecretsChangedEventArgs e)
...
+                    if (currentRetry > _retryCount || !IsTransient(ex))
+                    {
+                        throw;
+                    }
+                }
+                await Task.Delay(_retryDelay);
+            }
+        }
+
+        private async Task<FunctionSecrets> LoadFunctionSecretsWithRetryAsync(string functionName)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Security/SecretManager.cs]: Can we avoid the duplication here by factoring the common fault handling code into a method that takes a delegate?

File: src/WebJobs.Script.WebHost/Security/SecretManager.cs [hunk 4]
```diff
@@ -335,7 +337,23 @@ private Task<ScriptSecrets> LoadSecretsAsync(ScriptSecretsType secretsType, stri
...
-            string secretsJson = await _repository.ReadAsync(type, keyScope).ConfigureAwait(false);
+            string secretsJson;
+            for (int currentRetry = 0; ; currentRetry++)
+            {
+                try
+                {
+                    secretsJson = await _repository.ReadAsync(type, keyScope).ConfigureAwait(false);
+                    break;
+                }
+                catch (IOException)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Security/SecretManager.cs]: This class generalizes the storage repository (e.g. file based and blob based repositories). However, IOException is file specific? I don't think we should have repository specific logic in this general class.

Is the issue only happening for file storage? Perhaps we should push this file based retry logic further down into the FileUtility?

Comment (alrod) [reply src/WebJobs.Script.WebHost/Security/SecretManager.cs]: The idea was to make this change on SecurityManager level so we'll have retry logic for both (FileSystem and BlobStorage). Then we can just add new exception type we want to catch for retry.

Comment (fabiocav) [reply src/WebJobs.Script.WebHost/Security/SecretManager.cs]: I agree. As previously discussed, we want to keep any notion of file access/IO outside of this class as it should make not assumption about how the secrets are persisted (or if they are persisted).

**Comment by fabiocav** [comment]
@alrod , just wanted to make sure. This hasn't been iterated on since the last round of feedback. Do you have changes that have not yet been pushed?

**Pull request merged**